src/python/createrepo_c/__init__.py

```diff
@@ -382,6 +386,364 @@ def __init__(self, primary_path, filelists_path, other_path, newpkgcb=None, warn
+
+    @staticmethod
+    def from_path(path, warningcb=None):
+        """Construct a parser from an on-disk repository."""
+
+        if not warningcb:
+            def _warningcb(warning_type, message):
+                print("PARSER WARNING: %s" % message)
+                return True
+            warningcb = _warningcb
+
+        repomd_path = Path(path) / "repodata" / "repomd.xml"
+        if not repomd_path.exists():
+            raise FileNotFoundError("No repository found at the provided path.")
+
+        repomd = Repomd(str(repomd_path))
+        metadata_files = {record.type: record for record in repomd.records}
+        parser = RepositoryReader()
+        parser._primary_xml_path = os.path.join(path, metadata_files["primary"].location_href)
+        parser._filelists_xml_path = os.path.join(path, metadata_files["filelists"].location_href)
+        parser._other_xml_path = os.path.join(path, metadata_files["other"].location_href)
+
+        if metadata_files.get("updateinfo"):
+            parser._updateinfo_path = os.path.join(path, metadata_files["updateinfo"].location_href)
+        parser.repomd = repomd
+        return parser
+
+    @staticmethod
+    def from_metadata_files(primary_xml_path, filelists_xml_path, other_xml_path, updateinfo_xml_path=None):
+        """Construct a parser from the three main metadata files."""
+
+        parser = RepositoryReader()
+        parser._primary_xml_path = primary_xml_path
+        parser._filelists_xml_path = filelists_xml_path
+        parser._other_xml_path = other_xml_path
+        parser._updateinfo_path = updateinfo_xml_path
+
+        return parser
+
+    def advisories(self):
+        """Get advisories"""
+        if not self._updateinfo_path:
+            return []
+        else:
+            return UpdateInfo(self._updateinfo_path).updates
+
+    def package_count(self):
+        """Count the total number of packages."""
+        # It would be much faster to just read the number in the header of the metadata.
+        # But there's no way to do that. This gets fuzzy around the topic of duplicates.
+        # If the same package is listed more than once, is that counted as more than one package?
+        # Currently, no.
+        return len(self.parse_packages(only_primary=True)[0])
+
+    def iter_packages(self, warningcb=None):
+        """
+        Return an object which permits iterating over packages one at a time.
+
+        This uses less memory than parse_packages() and works in the presence of duplicate packages.
+
+        Kwargs:
+            warningcb (callable): A callback function for warnings emitted by the parser.
+
+        Returns:
+            cr.PackageIterator object.
+        """
+        return PackageIterator(
+            primary_path=self._primary_xml_path,
+            filelists_path=self._filelists_xml_path,
+            other_path=self._other_xml_path,
+            warningcb=warningcb,
+        )
+
+    def parse_packages(self, only_primary=False):
```

Comment:
I occasionally find this useful, but I'm not sure if it's useful enough to add to the API.  Thoughts?

Reply:
I think this is quite useful, especially for repodata introspection.